### DISCUSSION TITLE
Add `searchParams` to opengraph-image.tsx to allow for better OG image generation

### DISCUSSION THREAD

**Discussion by leonlarsson** [vercel/next.js#56314] [App Router]
As the title says, I would love to be able to access `searchParams` inside opengraph-image.tsx files.
This would allow you to generate a OpenGraph images based on `searchParams` (and therefore reflect URL state in the image).

In a page.tsx, this works perfectly well:
```tsx
// page.tsx
type Props = {
  searchParams: {
    search?: string;
    sort?: string;
  };
};

export default ({ searchParams }: Props) => { ... }
```

Doing the same in opengraph-image.tsx leads to `searchParams` being undefined.

As written in the documentation (https://nextjs.org/docs/app/api-reference/file-conventions/metadata/opengraph-image#props), it only has access to the `params`, which is not enough in my opinion.

Let's say my page relied on storing state in the URL for shareability.
I would need opengraph-image.tsx to be able to access `input` here: `example.com/?input=hello`

Thanks, and I hope this will be considered, or that I am outdated and it actually works somehow :)

**Comment by leonlarsson** [comment]
Quarterly bump :)

**Comment by leonlarsson** [comment]
One solution is to create a route (let's say `/og`) that generates and returns an image. Then we can simply link to the API in our metadata and it works, as we can pass searchParams to the route URL.

It would still be great UX to be able to do this directly in opengraph-image.tsx

**Comment by pjsny** [comment]
Bump- would be so useful

**Comment by carledwardfp** [comment]
bump

**Comment by hatchli** [comment]
Bump

**Comment by mc-cari** [comment]
Bump

**Comment by jonyx225** [comment]
bump

**Comment by minajevs** [comment]
Time wasted on debugging this issue and looking for solution online: 
1 hour


Feel free to increment it further :) 




<details><summary></summary>
<p>
this kind of issue is the reason for nextjs fatigue. production-ready my ass
</p>
</details>

**Comment by fcFn** [comment]
Adding the workaround in the docs is a good start. The workaround being having a dedicated route for handling image requests.

**Comment by superayan** [comment]
Bump!

**Comment by hWonderOfTheWorld** [comment]
bump

**Comment by Cary0x** [comment]
Dynamic og generation is a great idea. +1

**Comment by peetzweg** [comment]
Not a nextjs expert but while trying to todo this, the following guide with example was the only thing which was working for me as expected with the most recent nextjs 15.

https://vercel.com/guides/dynamic-text-as-image

**Comment by horainuri** [comment]
Bump

**Comment by livetOwO** [comment]
Bump

**Comment by wynandfyve** [comment]
Bump

**Comment by jcoulaud** [comment]
bump

**Comment by GeorgeKolokotronis** [comment]
bump

**Comment by pjsny** [comment]
bump

**Comment by shawngustaw** [comment]
Kind of insane that 2 years later this still isn't possible.

**Comment by carlydotwebsite** [comment]
ah, heck. well, consider this one more upvote, haha

**Comment by RustyNova016** [comment]
2026 bump!
